### PR TITLE
tools/mkimage: initialize dumpimage magic

### LINKED ISSUE #25287: tools/mkimage: dumpimage rejects valid legacy uImages after custom magic patch
### Describe the bug

Hi!

While working with the OpenWrt image tools I noticed that the OpenWrt-built `dumpimage` rejects valid legacy uImages, even though the matching `mkimage` binary reads the same image correctly

I tracked this down to `tools/mkimage/patches/030-allow-to-use-different-magic.patch`

The patch adds a `magic` field to `struct image_tool_params` and changes the legacy uImage check in `tools/default_image.c` from:

```c
if (be32_to_cpu(hdr->ih_magic) != IH_MAGIC)
```

to:

```c
if (be32_to_cpu(hdr->ih_magic) != params->magic)
```

`mkimage.c` initializes it correctly:

```c
static struct image_tool_params params = {
    ...
    .magic = IH_MAGIC,
};
```

but `dumpimage.c` still has:

```c
static struct image_tool_params params;
```

so `params.magic` stays zero

### OpenWrt version

r36385-c1111ed5d6

### OpenWrt release

SNAPSHOT

### OpenWrt target/subtarget

ramips/mt7621

### Device

TP-Link Archer MR500 v1 (EU)

### Image kind

Self-built image

### Steps to reproduce

I noticed this while testing the OpenWrt host image tools on a locally built `ramips/mt7621` image.

The extracted kernel is a valid legacy uImage. Its magic, header CRC and data CRC were also verified independently.

Using the OpenWrt-built U-Boot 2026.07 host tools:

```sh
mkimage -l kernel.bin
mkimage -T kernel -l kernel.bin
```

Both commands succeed

Then run:

```sh
dumpimage -l kernel.bin
```

and:

```sh
dumpimage -T kernel -l kernel.bin
```

The issue is reproducible after a clean rebuild of the OpenWrt `tools/mkimage` host package as well

### Actual behaviour

`mkimage` accepts the image and reports:

```text
Image Name:   MIPS OpenWrt Linux-6.18.52
Image Type:   MIPS Linux Kernel Image (lzma compressed)
Data Size:    3561279 Bytes
Load Address: 80001000
Entry Point:  80001000
```

However, `dumpimage -l kernel.bin` fails with:

```text
Truncated file
dumpimage: cannot detect image type
```

and forcing the image type:

```sh
dumpimage -T kernel -l kernel.bin
```

fails with:

```text
dumpimage: verify_header failed for Default Image support with exit code -9
```

`-9` corresponds to `FDT_ERR_BADMAGIC`

### Expected behaviour

`dumpimage` should accept the same valid legacy uImage that `mkimage` accepts

It should also be possible to extract the payload successfully with:

```sh
dumpimage -T kernel -p 0 -o payload kernel.bin
```

### Additional info

Tested with:

```text
OpenWrt revision: c1111ed5d69678613b7a47aa76f8e6209f0da56d
U-Boot host tools: 2026.07
Host: x86_64 Linux
```

The test image was independently checked and has:

```text
magic:       0x27051956
header CRC:  valid
data CRC:    valid
payload:     3561279 bytes
```

I also verified that the kernel extracted from the firmware is byte-for-byte identical to the kernel artifact produced by the OpenWrt build.

The issue was reproduced again after cleaning and rebuilding the OpenWrt U-Boot host tools, so it does not appear to be caused by a stale binary.

The apparent cause is the OpenWrt custom-magic patch.

`tools/default_image.c` checks:

```c
if (be32_to_cpu(hdr->ih_magic) != params->magic)
```

but `dumpimage.c` leaves `params.magic` uninitialized, while `mkimage.c` initializes it to `IH_MAGIC`.

As a test, I changed the declaration in `dumpimage.c` to:

```c
static struct image_tool_params params = {
    .magic = IH_MAGIC,
};
```

and rebuilt the same U-Boot host tools.

After that, all of these commands succeeded:

```text
dumpimage -T kernel -l kernel.bin                -> OK
dumpimage -T kernel -p 0 -o payload kernel.bin  -> OK
dumpimage -l kernel.bin                          -> OK
```

The extracted payload was exactly `3561279` bytes, matching the uImage header.

After testing, I restored the original source and rebuilt again, confirming that the original `dumpimage` behaviour returned

### Diffconfig

```text
CONFIG_TARGET_ramips=y
CONFIG_TARGET_ramips_mt7621=y
CONFIG_TARGET_ramips_mt7621_DEVICE_tplink_mr500-v1-eu=y
CONFIG_DEVEL=y
CONFIG_CCACHE=y
CONFIG_CCACHE_DIR="/home/max/.cache/openwrt-ccache"
CONFIG_DEFAULT_kmod-usb-net-rndis=y
# CONFIG_FEED_luci is not set
# CONFIG_FEED_packages is not set
# CONFIG_FEED_routing is not set
# CONFIG_FEED_telephony is not set
# CONFIG_FEED_video is not set
CONFIG_PACKAGE_kmod-usb-net-rndis=y
```

### Terms

- [x] I am reporting an issue for OpenWrt, not an unsupported fork.

### PR DESCRIPTION
**Pull request by massimomazzariol**
OpenWrt's custom uImage magic patch changes the legacy image verifier to use params.magic instead of IH_MAGIC.

mkimage initializes this field, but dumpimage does not, leaving it zeroed and causing valid legacy uImages to fail verification with FDT_ERR_BADMAGIC.

Initialize dumpimage with the normal IH_MAGIC, matching mkimage.

Tested with the OpenWrt-built U-Boot 2026.07 host tools:

    dumpimage -T kernel -l kernel.bin                -> OK
    dumpimage -T kernel -p 0 -o payload kernel.bin  -> OK
    dumpimage -l kernel.bin                          -> OK

The extracted payload is 3561279 bytes, matching the legacy uImage header.

Fixes #25287

### CODE REVIEW

File: tools/mkimage/patches/030-allow-to-use-different-magic.patch [hunk 1]
```diff
@@ -5,6 +5,19 @@ Subject: [PATCH] tools/mkimage: allow to use different magic in uImages
 
 This patch makes it possible to set a custom image magic.
 
+--- a/tools/dumpimage.c
++++ b/tools/dumpimage.c
+@@ -12,7 +12,9 @@
+ static void usage(void);
+
```
Comment (openwrt-ai) [inline tools/mkimage/patches/030-allow-to-use-different-magic.patch]: nit: this blank context line, and the one on line 18, are fully empty instead of a single-space line; every other hunk in this file keeps the leading space on empty context lines (e.g. [the blanks around `static const char optstring[]`](``https://github.com/openwrt/openwrt/blob/da5334f03700b63a5b59291eb4208149eb4e794c/tools/mkimage/patches/030-allow-to-use-different-magic.patch#L47-L52)).`` GNU patch tolerates it, so nothing breaks today, but the next patch refresh will rewrite both lines and add unrelated churn. Restore the leading space on lines 12 and 18.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/mkimage/patches/030-allow-to-use-different-magic.patch [hunk 2]
```diff
@@ -5,6 +5,19 @@ Subject: [PATCH] tools/mkimage: allow to use different magic in uImages
...
+ /* parameters initialized by core will be used by the image type code */
+-static struct image_tool_params params;
++static struct image_tool_params params = {
++	.magic = IH_MAGIC,
```
Comment (openwrt-ai) [inline tools/mkimage/patches/030-allow-to-use-different-magic.patch:16]: nit: `dumpimage` still has no `-M` handling in its own `process_args()`, so after this it can only ever read standard-magic images — the many in-tree kernels built with a custom magic, e.g. [`uImage lzma -M 0x4f4b4c49` in ath79 generic.mk](https://github.com/openwrt/openwrt/blob/da5334f03700b63a5b59291eb4208149eb4e794c/target/linux/ath79/image/generic.mk#L399), stay undumpable. Was mirroring [the `case 'M':` block this patch adds to mkimage](https://github.com/openwrt/openwrt/blob/da5334f03700b63a5b59291eb4208149eb4e794c/tools/mkimage/patches/030-allow-to-use-different-magic.patch#L60-L68) into dumpimage considered, or is restoring the default-magic path deliberately the whole scope here?

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/mkimage/patches/030-allow-to-use-different-magic.patch (hunk 1 above)
Comment (massimomazzariol) [reply tools/mkimage/patches/030-allow-to-use-different-magic.patch]: It should be fixed by refreshing the patch, now it applies cleanly without offset or fuzz

File: tools/mkimage/patches/030-allow-to-use-different-magic.patch (hunk 2 above)
Comment (massimomazzariol) [reply tools/mkimage/patches/030-allow-to-use-different-magic.patch:16]: I added matching -M handling to dumpimage and updated the usage text. I also tested both the default IH_MAGIC path and a custom image, including extraction and negative cases

**Review by openwrt-ai — COMMENTED** [review]
Reviewed 1 new commit; no new issues found.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by robimarko** [comment]
Does it check if the value is outside of 32 bit uImage magic range?

If not, probably should check for that

**Comment by massimomazzariol** [comment]
> Does it check if the value is outside of 32 bit uImage magic range?
> 
> If not, probably should check for that

Thanks for the osservation @robimarko now the fix is applied

`dumpimage` now rejects values outside the 32-bit uImage magic range before assigning them to `params.magic`

I reproduced the previous truncation behaviour and verified the boundaries: `0xffffffff` is accepted while `0x100000000` and larger values are rejected. The default and custom magic paths still pass, including extraction tests

I also noticed that the existing `mkimage -M` handling has the same unchecked narrowing. Would it make sense to address that separately in a follow-up PR, so this PR stays focused on `dumpimage` or should I keep everything here? What do you think?

**Comment by robimarko** [comment]
Please adress it in the same PR, just in additional commit, thanks

**Comment by massimomazzariol** [comment]
> Please adress it in the same PR, just in additional commit, thanks

Ready for review. Thanks!
